[RLlib] DreamerV3 on PyTorch.

sven1977 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

DreamerV3 on PyTorch.

The RLlib team finally managed to fully transfer DreamerV3 to PyTorch, including support for multi-GPU training.

Training (and sampling) throughput nicely scale linearly now with the number of GPUs/Learners used:

<img width="434" height="305" alt="image" src="https://github.com/user-attachments/assets/9f3973b1-371d-415b-9c16-6db6d7996a05" />

Learning speed should also scale, albeit most likely sub-linearly. Learning rate adjustments are crucial to get right. Below is a 4GPU run with learning rates 4x and a 2GPU run with learning rates 1.4x (x2**0.5).
<img width="436" height="309" alt="image" src="https://github.com/user-attachments/assets/f81c4bc6-05c8-4e3d-b448-f0d569262b60" />

For 8 GPUs or more linear scaling for Pong becomes more difficult (but this may not apply to harder envs), but more complex lr-schedules may help.


<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into upgrade_gymnasium_to_1_0_0a1

sven1977 committed:
LINT

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
LINT

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Apply suggestions from code review

Signed-off-by: Sven Mika <sven@anyscale.io>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge remote-tracking branch 'origin/upgrade_gymnasium_to_1_0_0a1' into upgrade_gymnasium_to_1_0_0a1

# Conflicts:
#	rllib/env/single_agent_env_runner.py

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into upgrade_gymnasium_to_1_0_0a1

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

Signed-off-by: sven1977 <svenmika1977@gmail.com>

# Conflicts:
#	rllib/BUILD
#	rllib/algorithms/algorithm.py
#	rllib/algorithms/dreamerv3/dreamerv3_rl_module.py
#	rllib/algorithms/dreamerv3/tests/test_dreamerv3.py
#	rllib/algorithms/dreamerv3/tf/dreamerv3_tf_rl_module.py
#	rllib/algorithms/dreamerv3/utils/env_runner.py
#	rllib/core/rl_module/torch/torch_rl_module.py
#	rllib/utils/tests/test_torch_utils.py

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
torch compilation tests running (Atari not tested due to Mac error)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
LINT

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
compilation tests run again, even with latest pytorch version, which was responsible for breaking a test

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
GPU on EnvRunner fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
GPU in Stride2D fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
run now on 1 GPU (local learner) w/ shared EnvRunner/Learner RLModule and Atari (Pong) on a V100 machine. :+1:

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
DM Control Suite example

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixed separation issue of actor/critic losses messing with the world model grads

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
tuned_examples cleanup

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 commented on rllib/BUILD:
```diff
@@ -1482,6 +1482,13 @@ py_test(
     srcs = ["utils/exploration/tests/test_random_encoder.py"]
 )
 
+py_test(
```
added `two_hot` tests (for torch, too, of course).

sven1977 commented on rllib/algorithms/dreamerv3/dreamerv3.py:
```diff
@@ -28,24 +26,29 @@
 from ray.rllib.core import DEFAULT_MODULE_ID
 from ray.rllib.core.columns import Columns
 from ray.rllib.core.rl_module.rl_module import SingleAgentRLModuleSpec
+from ray.rllib.execution.rollout_ops import synchronous_parallel_sample
```
This is the most important file to review, the rest is mere "translation" from the existing tf version to torch.

sven1977 committed:
cleanup

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 commented on rllib/algorithms/dreamerv3/utils/summaries.py:
```diff
@@ -115,7 +116,7 @@ def report_dreamed_trajectory(
 def report_predicted_vs_sampled_obs(
     *,
-    results,
+    metrics,
```
We pass in the main Algo's metrics object now and just log into that (instead of returning altered dicts).

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -1045,36 +1041,13 @@ def additional_update_for_module(
         self,
         batch: MultiAgentBatch,
         *,
+        timesteps=None,  # TODO
```
this is still a TODO for this PR:
* We should pass in the current global timesteps (env, agents and modules) into all `update` calls
* instead, we skip the `additional_update` call and replace it with a user-overridable before_update and after_update sub-method, called automatically by `Learner.update_from_batch|episodes`.
* In the future, we might simplify by deprecating `update_from_batch` and fully replace it with `update`, which will take only episodes. ??

simonsays1980 replied:
As we fully focus on episodes this makes sense in my opinion.

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -1388,22 +1356,36 @@ def _update_from_batch_or_episodes(
-            },
-            window=1,
-        )
+        self._after_update(timesteps)
```
maybe make this public

sven1977 commented on rllib/core/models/torch/utils.py:
```diff
@@ -45,28 +45,33 @@ def __init__(self, width, height, stride_w, stride_h):
-            size=(
-                self.width * self.stride_w - (self.stride_w - 1),
-                self.height * self.stride_h - (self.stride_h - 1),
+        self.register_buffer(
```
make this GPU-capable.

sven1977 committed:
fix critic loss (the main reduction has to be mean instead of sum)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

simonsays1980 commented on rllib/core/learner/learner.py:
```diff
@@ -1388,22 +1356,36 @@ def _update_from_batch_or_episodes(
 
+    # TODO (sven): Document and design this new approach in order to replace
+    #  `additional_update()`.
+    def _after_update(self, timesteps):
```
I thought this function needs to be public, but we are using also `_forward_train`, etc. for overriding in the `RLModule` .... so this should be fine.

simonsays1980 commented on rllib/tuned_examples/dreamerv3/atari_200M.py:
```diff
@@ -50,10 +50,21 @@
+        remote_worker_envs=True,
+    )
+    .learners(
+        num_learners=0 if args.num_gpus == 1 else args.num_gpus,
```
Btw, isn't it possible to use here fractional GPUs as well? Even though the model is large I guess a V100 could be shared between 2 learners?

simonsays1980 commented on rllib/utils/torch_utils.py:
```diff
@@ -684,6 +684,129 @@ def softmax_cross_entropy_with_logits(
 
 
+@PublicAPI
+def symlog(x: "torch.Tensor") -> "torch.Tensor":
```
Can't we compile such torch functions?

simonsays1980 commented on rllib/utils/torch_utils.py:
```diff
@@ -684,6 +684,129 @@ def softmax_cross_entropy_with_logits(
+
 def _dynamo_is_available():
     # This only works if torch._dynamo is available
     try:
```
Btw is this now available on CI?

simonsays1980 reviewed (approved):
LGTM. Awesome PR. So great to have it now in PyTorch!

sven1977 committed:
add support for reporting posterior predicted (decoder) videos

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
less frequent video logging to save time (very expensive)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix: less frequent video logging to save time (very expensive)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix: less frequent video logging to save time (very expensive)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
adjust initializers to match 100% those of Author's and RLlib's tf versions.

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
add independent wrapper around one-hot-categorical to representation layer's (32x32 z state) distribution.

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
support dream data summaries comparing dreamed trajectories (from prior states and w/ videos) to the sampled ones.

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
enable dream data video/reports.

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
manual GRU

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
GPU fix

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
move even more toward JAX repo by initializing all layers via Danijar's almost-glorot-normal initializer.

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix dreamed reports (by reducing to B=1, which seems to work with WandB; could be that WandB doesn't like lists of lists).

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
small cleanup

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
Big fix in DreamerV3 EnvRunner: States were not properly passed around between calls to sample()

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
cleanup

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

Signed-off-by: sven1977 <svenmika1977@gmail.com>

# Conflicts:
#	rllib/utils/metrics/metrics_logger.py

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into upgrade_gymnasium_to_1_0_0a1

Signed-off-by: sven1977 <svenmika1977@gmail.com>

# Conflicts:
#	rllib/env/single_agent_env_runner.py

sven1977 committed:
Update to gymnasium 1.0.0a2 and new ale_py (which now supports Atari)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Update to gymnasium 1.0.0a2 and new ale_py (which now supports Atari)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
merge

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 commented on rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:
```diff
@@ -0,0 +1,490 @@
+        An initial state is generated using the previous action, the tanh of the
+        (learned) h-state variable and the dynamics predictor (or "prior net") to
+        compute z^0 from h0. In this last step, it is important that we do NOT sample
+        the z^-state (as we would usually do during dreaming), but rather take the mode
```
Z-hat (z^) is the "prior" z-state, meaning a z-state that gets computed using only(!) the h-state as input.
On the other hand, a z-state (without the hat) gets computed using the "posterior" net, which has access to the h-state AND the encoder's output.

sven1977 commented on rllib/algorithms/dreamerv3/torch/models/components/representation_layer.py:
```diff
@@ -0,0 +1,128 @@
+        distribution = torch.distributions.OneHotCategorical(logits=logits)
+
+        # Draw a one-hot sample (B, num_categoricals, num_classes).
+        sample = distribution.sample()
```
I guess, we could do `rsample`, but I felt it was much safer to do the straight-through sampling explicitly here in the code.

sven1977 commented on rllib/algorithms/dreamerv3/torch/dreamerv3_torch_learner.py:
```diff
@@ -0,0 +1,913 @@
+    The critic EMA-copy update step can be found in the `DreamerV3Learner` base class,
+    as it is framework independent.
+
+    We define 3 local TensorFlow optimizers for the sub components "world_model",
```
great catch. Fixed.

sven1977 commented on rllib/algorithms/dreamerv3/tests/test_dreamerv3.py:
```diff
@@ -209,32 +242,231 @@ def test_dreamerv3_dreamer_model_sizes(self):
+                modules = []
+                tf_weights = []
+
+                for fw in framework_iterator(config, frameworks=("tf2", "torch")):
```
We will, but first, we need to make the torch version learn Atari :) 
I'll remove these tests once this is all confirmed and then we can remove the tf2 stuff as well.

sven1977 commented on rllib/algorithms/dreamerv3/dreamerv3.py:
```diff
@@ -631,71 +671,50 @@ def training_step(self) -> ResultDict:
 
         # Return all results.
-        return results
+        return self.metrics.reduce()
```
In the future, we should reduce later (users should NOT call reduce() anymore at all, anywhere). But right now, we still require this here. Should be a separate PR, imo.

sven1977 commented on rllib/algorithms/dreamerv3/dreamerv3.py:
```diff
@@ -535,45 +544,74 @@ def training_step(self) -> ResultDict:
-                if self._counters[NUM_AGENT_STEPS_SAMPLED] == 0:
-                    d_, o_ = env_runner.sample(
-                        num_timesteps=(
+                if self.metrics.peek(NUM_ENV_STEPS_SAMPLED_LIFETIME, default=0) == 0:
```
Ah, great point. I think you are right. ...

sven1977 commented on rllib/algorithms/dreamerv3/dreamerv3.py:
```diff
@@ -506,8 +517,6 @@ def setup(self, config: AlgorithmConfig):
     def training_step(self) -> ResultDict:
-        results = {}
-
         env_runner = self.workers.local_worker()
```
done

sven1977 commented on rllib/algorithms/dreamerv3/dreamerv3.py:
```diff
@@ -535,45 +544,74 @@ def training_step(self) -> ResultDict:
             ):
                 # Sample using the env runner's module.
-                done_episodes, ongoing_episodes = env_runner.sample()
+                episodes, env_runner_results = synchronous_parallel_sample(
```
see further below: This property always computes the actual current training ratio using the actual counts.
```
@property
def training_ratio(self) -> float:

```

sven1977 commented on rllib/algorithms/dreamerv3/dreamerv3.py:
```diff
@@ -421,7 +420,13 @@ def validate(self) -> None:
     @override(AlgorithmConfig)
     def get_default_learner_class(self):
-        if self.framework_str == "tf2":
+        if self.framework_str == "torch":
```
I wish! Torch needs to learn Atari first. We are close, but it's not learning yet. Must find the "final bug" before we can merge this ...

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
LINT

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
torch Stats just like tf Stats to try whether the torch.nanmean() on the Stats.values list is the problem ... to be checked

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
merge

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

Signed-off-by: sven1977 <svenmika1977@gmail.com>

# Conflicts:
#	rllib/utils/metrics/stats.py

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into upgrade_gymnasium_to_1_0_0a1

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into upgrade_gymnasium_to_1_0_0a1

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into appo_on_new_api_stack_w_env_runner_and_connectorv2

Signed-off-by: sven1977 <svenmika1977@gmail.com>

# Conflicts:
#	rllib/env/env_runner_group.py

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into upgrade_gymnasium_to_1_0_0a1

Signed-off-by: sven1977 <svenmika1977@gmail.com>

# Conflicts:
#	rllib/examples/connectors/frame_stacking.py

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
Merge branch 'upgrade_gymnasium_to_1_0_0a1' into dreamer_v3_torch

sven1977 committed:
merged in gymnasium=1.0.0 PR to unify dreamer EnvRunner with SingleAgentEnvRunner (and deprecate dreamer's env-runner).

Will test this with torch, then apply the changes to the current tf-Dreamer PR (meaning the gymnasium==1.0.0 PR, which stil only has dreamer+Tf support).

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

Viibrant commented:
Hi all, what is the current status of this PR?

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

Signed-off-by: sven1977 <svenmika1977@gmail.com>

# Conflicts:
#	doc/source/rllib/doc_code/training.py
#	doc/source/rllib/rllib-examples.rst
#	python/requirements.txt
#	python/requirements/ml/rllib-test-requirements.txt
#	python/requirements_compiled.txt
#	python/setup.py
#	release/ray_release/byod/requirements_byod_3.9.txt
#	rllib/algorithms/algorithm.py
#	rllib/algorithms/algorithm_config.py
#	rllib/algorithms/bc/bc.py
#	rllib/algorithms/dreamerv3/dreamerv3.py
#	rllib/algorithms/dreamerv3/dreamerv3_rl_module.py
#	rllib/algorithms/dreamerv3/tests/test_dreamerv3.py
#	rllib/algorithms/dreamerv3/utils/env_runner.py
#	rllib/algorithms/ppo/ppo.py
#	rllib/algorithms/ppo/tests/test_ppo.py
#	rllib/algorithms/ppo/tests/test_ppo_rl_module.py
#	rllib/algorithms/ppo/tests/test_ppo_with_env_runner.py
#	rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py
#	rllib/benchmarks/torch_compile/run_inference_bm.py
#	rllib/benchmarks/torch_compile/run_ppo_with_inference_bm.py
#	rllib/connectors/common/add_states_from_episodes_to_batch.py
#	rllib/core/models/specs/tests/test_tensor_spec.py
#	rllib/core/models/tests/test_cnn_transpose_heads.py
#	rllib/env/single_agent_env_runner.py
#	rllib/env/single_agent_episode.py
#	rllib/env/tests/test_single_agent_env_runner.py
#	rllib/env/utils/__init__.py
#	rllib/examples/_old_api_stack/custom_keras_model.py
#	rllib/examples/envs/env_rendering_and_recording.py
#	rllib/examples/rl_modules/custom_cnn_rl_module.py
#	rllib/models/tests/test_preprocessors.py
#	rllib/utils/metrics/metrics_logger.py
#	rllib/utils/metrics/stats.py
#	rllib/utils/tests/test_tf_utils.py

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
wip

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix >1 GPU problem on torch

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes for increased LRs for >1 learners

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixed custom GRU layer and setup (layernorm was missing within GRU, also no bias on the custom GRU Linear layer!)

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
remove tf comparison code

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into dreamer_v3_torch

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fixes

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix

Signed-off-by: sven1977 <svenmika1977@gmail.com>

sven1977 committed:
fix inference logic (broken due to time rank in returned actions and missing batch in init. state)
fix bug in forward inference: use `dist.mode` instead of `dist.mode()`
add no_grad around forward_inference and _exploration in DreamerV3 RLModule.

Signed-off-by: sven1977 <svenmika1977@gmail.com>
